src/simoc_sam/mqttbridge.py

```diff
@@ -0,0 +1,276 @@
+    """Forward the refresh-sensor call to one or all sensor managers"""
+    print('Refreshing sensors (from server)')
+    for sm_id in SENSOR_MANAGERS:
+        if sensor_manager_id is None or sm_id == sensor_manager_id:
+            await sio.emit('refresh-sensors', to=sm_id)
+
+def get_timestamp():
+    """Return the current timestamp as a string."""
+    return datetime.now().strftime('%Y-%m-%d %H:%M:%S')
+
+
+# main loop that broadcasts bundles
+
+async def emit_readings():
+    """Emit a bundle with the latest reading of all sensors."""
+    args = utils.parse_args()  # TODO: create separate parser for the server
+    delay = args.delay
+    print(f'Broadcasting data every {delay} seconds.')
+    n = 0
+    while True:
+        # TODO: set up a room for the clients and broadcast to the room
+        # TODO: improve ctrl+c handling (see graceful shutdown)
+        timestamp = get_timestamp()
+        print('SR', SENSOR_READINGS)
+        sensors_readings = {sid: readings[-1]
+                            for sid, readings in SENSOR_READINGS.items()
+                            if readings}
+        bundle = dict(n=n, timestamp=timestamp, readings=sensors_readings)
+        # print(bundle)
+        try:
+            # the frontend expects a list of bundles
+            await emit_to_subscribers('step-batch', [bundle])
+            n += 1
+            print(f'{len(SENSORS)} sensor(s); {len(SUBSCRIBERS)} '
+                    f'subscriber(s); {n} readings broadcasted')
+        except Exception as e:
+            print('!!! Failed to emit step-batch:')
+            traceback.print_exc()
+        await sio.sleep(delay)
+
+
+async def mqtt_handler():
+    broker_address = "samrpi1"
+    client = aiomqtt.Client(broker_address)
+    print(SENSOR_INFO)
+    interval = 5  # Seconds
+    while True:
+        try:
+            async with client:
+                await client.subscribe("sam/#")
+                async for message in client.messages:
+                    topic = message.topic.value
+                    print(topic)
+                    sam, host, sensor = topic.split('/')
+                    sensor_id = f'{host}.{sensor}'
+                    if sensor_id not in SENSOR_INFO:
+                        info = copy.deepcopy(SENSOR_DATA[sensor])
+                        info['sensor_name'] = sensor_id
+                        info['sensor_id'] = sensor_id
+                        info['sensor_desc'] = f'{sensor} sensor on {host}'
+                        SENSOR_INFO[sensor_id] = info
+                        await emit_to_subscribers('sensor-info', SENSOR_INFO)
+
+                    payload = json.loads(message.payload.decode())
+                    # print('adding', payload)
+                    SENSOR_READINGS[sensor_id].append(payload)
+        except aiomqtt.MqttError as err:
+            print(f'Connection lost; {err}')
+            print(f'Reconnecting in {interval} seconds...')
+            await asyncio.sleep(interval)
```

Comment:
This code is copied from this example: https://sbtinstruments.github.io/aiomqtt/reconnection.html
However sometimes it gives an error saying that the cm is not reentrant.
Maybe we could move the client instantiation inside the `while True:`

Reply:
I moved the client instantiation inside the `while True:`